File: client-android/lib/src/main/kotlin/stasis/client_android/lib/tracking/state/serdes/BackupStateSerdes.kt
package stasis.client_android.lib.tracking.state.serdes

import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.persistence.state.StateStore
import stasis.client_android.lib.tracking.state.BackupState
import stasis.client_android.lib.utils.Try
import java.util.UUID

object BackupStateSerdes : StateStore.Serdes<Map<OperationId, BackupState>> {
    override fun serialize(state: Map<OperationId, BackupState>): ByteArray =
        stasis.client_android.lib.model.proto.BackupStateCollection(
            collection = state.map { (k, v) -> k.toString() to BackupState.toProto(state = v) }.toMap()
        ).encode()

    override fun deserialize(bytes: ByteArray): Try<Map<OperationId, BackupState>> = Try {
        stasis.client_android.lib.model.proto.BackupStateCollection.ADAPTER
            .decode(bytes)
            .collection
            .map { (k, v) ->
                val operation = UUID.fromString(k)
                operation to BackupState.fromProto(operation = operation, state = v).get()
            }.toMap()
    }
}

File: client/src/main/scala/stasis/client/tracking/state/serdes/BackupStateSerdes.scala
package stasis.client.tracking.state.serdes

import java.util.UUID

import scala.util.Try

import stasis.client.model.proto
import stasis.client.tracking.state.BackupState
import stasis.core.persistence.backends.file.state.StateStore
import stasis.shared.ops.Operation

object BackupStateSerdes {
  implicit val backupSerdes: StateStore.Serdes[Map[Operation.Id, BackupState]] =
    new StateStore.Serdes[Map[Operation.Id, BackupState]] {
      override def serialize(state: Map[Operation.Id, BackupState]): Array[Byte] =
        proto.state
          .BackupStateCollection(
            collection = state.map { case (k, v) => k.toString -> BackupState.toProto(state = v) }
          )
          .toByteArray

      override def deserialize(bytes: Array[Byte]): Try[Map[Operation.Id, BackupState]] =
        proto.state.BackupStateCollection
          .parseFrom(bytes)
          .collection
          .foldLeft(Try(Seq.empty[BackupState])) { case (collected, (k, v)) =>
            for {
              collected <- collected
              state <- BackupState.fromProto(operation = UUID.fromString(k), state = v)
            } yield {
              collected :+ state
            }
          }
          .map(_.map(state => state.operation -> state).toMap)
    }
}
